### ISSUE TITLE
Decoding test failure in 3.13.1

### ISSUE THREAD

**Issue by jamesjer** [gap-packages/guava#22]
I built the 3.13.1 release on an x86_64 machine running Linux, specifically Fedora 24.  One decoding test sometimes fails, in two different ways.  I've manage to catch both failure types while running the test manually:
```
gap> x:= Indeterminate( GF(2) );
x_1
gap> P:= x^3+x^2+1;
x_1^3+x_1^2+Z(2)^0
gap> C := GeneratorPolCode(P, 7, GF(2));
a cyclic [7,4,1..3]1 code defined by generator polynomial over GF(2)
gap> HasSpecialDecoder(C);
true
gap> c := Random(C);
[ 0 1 1 0 0 0 1 ]
gap> i := InformationWord(C, c);
[ 0 1 1 1 ]
gap> e := ErrorVector(Random([1..WordLength(C)]), C);
[ 0 0 0 1 0 0 0 ]
gap> j := Decode(C,c+e);
[ 0 1 1 0 ]
gap> (i=j);
false
gap> w := Decodeword(C,c+e);
[ 0 1 1 1 0 1 0 ]
gap> (c=w);
false
```
and:
```
gap> c := Random(C);
[ 1 0 0 0 1 0 1 ]
gap> i := InformationWord(C, c);
[ 1 0 1 1 ]
gap> e := ErrorVector(Random([1..WordLength(C)]), C);
[ 0 0 0 0 1 0 0 ]
gap> j := Decode(C,c+e);
Error, List Element: <list>[1] must have an assigned value in
  return 0 \* cc[1]; at /builddir/build/BUILD/pkg/guava/lib/decoders.gi:307 called from 
func( C[i] ) at /usr/lib/gap/lib/coll.gi:746 called from
List( [ 1 .. n - r ], function ( k )
      return 0 \* cc[1];
  end ) at /builddir/build/BUILD/pkg/guava/lib/decoders.gi:307 called from
SpecialDecoder( C )( C, c 
 ) at /builddir/build/BUILD/pkg/guava/lib/decoders.gi:70 called from
<function "unknown">( <arguments> )
 called from read-eval loop at line 45 of _stdin_
you can 'return;' after assigning a value
```

**Comment by osj1961** [comment]
Hi Jerry,


Thanks for your bug report (and especially for the thorough investigation of the two distinct failure modes!)  So far, I've verified that this bug was present in earlier guava versions at least as far back as 3.10 -- which is when I took over maintenance.


After a brief look at the code I see several things that aren't quite right...  I will try to have a fix in place shortly, and share it with you.  I'm working towards the 3.14 release this Summer and this is definitely something that needs to be repaired before the new release.


Best wishes,

Joe


Joseph E. Fields
Professor of Mathematics
Southern Connecticut State University
http://www.southernct.edu/~fields/
________________________________

**Comment by osj1961** [comment]
Fixed with dd08130
